lib/puppet/type/zfs_share.rb

```diff
@@ -0,0 +1,141 @@
+Puppet::Type.newtype(:zfs_share) do
+  @doc = "Manage zfs shares"
+
+  require 'puppet/parameter/boolean'
+
+  ensurable
+
+  newparam(:zfs_name) do
+    desc "Name of zfs file system"
+    isnamevar
+
+    validate do |value|
+      unless value =~ /(?!\/|\d)^\w.*/
+        raise ArgumentError , "%s is not a valid zfs path."
+      end
+    end
+  end
+
+  newparam(:zpool) do
+    desc "Name of the zfs parent"
+
+    validate do |value|
+      unless value =~ /(?!\/|\d)^\w.*/
+        raise ArgumentError , "%s is not a valid zpool."
+      end
+    end
+  end
+
+  newparam(:share_title) do
+    desc "Override the title of the share"
+
+    validate do |value|
+      if value =~ /\s/
+        raise ArgumentError , "#{value} is not a valid share title"
+      end
+    end
+  end
+
+  newparam(:allow_ip_read, :array_matching => :all) do
+    desc "IP allowed to access share"
+
+    validate do |value|
+      if value.kind_of?(String)
+        value = [value]
+      end
+      value.each do |v|
+        unless v =~ /^\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\/\d{2}|\*/
+          raise ArgumentError , "%s is not a valid IP for setting a zfs share."
+        end
+      end
+    end
+  end
+
+  newparam(:allow_ip_write, :array_matching => :all) do
+    desc "IP allowed to access share"
+
+    defaultto '127.0.0.1/32'
+    validate do |value|
+      if value.kind_of?(String)
+        value = [value]
+      end
+      value.each do |v|
+        unless v =~ /^\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\/\d{2}|\*/
+          raise ArgumentError , "%s is not a valid IP for setting a zfs share."
+        end
+      end
+    end
+  end
+
+  newparam(:protocol, :array_matching => :all) do
+    desc "Protocol for share"
+
+    defaultto 'nfs'
+    validate do |value|
+      if value.kind_of?(String)
+        value = [value]
+      end
+      value.each do |v|
+        unless v =~ /nfs|smb/
+          raise ArgumentError , "%s is not a supported protocol. Only nfs and smb are supported"
+        end
+      end
+    end
+  end
+
+  newparam(:security, :array_matching => :all) do
+    desc "Security for share"
+
+    defaultto 'none'
+    validate do |value|
+      if value.kind_of?(String)
+        value = [value]
+      end
+      value.each do |v|
+        unless v =~ /sys|default|none/
+          raise ArgumentError , "%s is not a valid security type"
+        end
+      end
+    end
+  end
+
+  newparam(:anon_user_id) do
+    desc "UID to access as when connecting anonymously"
+
+    validate do |value|
+      unless value =~ /\d+/
+        raise ArgumentError , "%s is not a valid uid"
+      end
+    end
+  end
+
+  newparam(:nosub, :boolean => true, :parent => Puppet::Parameter::Boolean) do
+    desc "Prevents clients from mounting subdirectories of shared directories"
+
+    validate do |value|
+      unless value == true or false
```

Comment:
Did you test the usage of the parameter without this validate block?  Did it accept any arbitrary value and attempt passing it to the provider?  If it does this can probably be solved using the `newvalues` method.